File: 0144-Binary-Tree-Preorder-Traversal/Java/Solution.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * @author youyusong
 * @date 2018/12/18
 */
public class Solution {

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    /**
     * 给定一个二叉树，返回它的 前序 遍历。
     * @param root
     * @return
     */
//    public List<Integer> preorderTraversal(TreeNode root) {
//        List<Integer> list = new ArrayList<>();
//
//        preorderTraversal(root, list);
//
//        return list;
//    }
//
//    private void preorderTraversal(TreeNode node, List<Integer> list) {
//        if (node != null) {
//            list.add(node.val);
//            preorderTraversal(node.left, list);
//            preorderTraversal(node.right, list);
//        }
//    }

    /**
     * 非递归实现
     * @param root
     * @return
     */
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        if (root == null) {
            return list;
        }

        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);

        while (! stack.isEmpty()) {
            TreeNode node = stack.pop();
            list.add(node.val);

            if (node.right != null) {
                stack.push(node.right);
            }

            if (node.left != null) {
                stack.push(node.left);
            }
        }

        return list;
    }

}

File: 0144-Binary-Tree-Preorder-Traversal/Python/Solution.py
# Definition for a binary tree node.
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:

    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result: List[int] = []
        self._preorder(root, result)
        return result


    def _preorder(self, node: Optional[TreeNode], result: List[int]) -> None:
        if node is None:
            return
        result.append(node.val)
        self._preorder(node.left, result)
        self._preorder(node.right, result)